libopenage/unit/research.cpp

```diff
@@ -0,0 +1,59 @@
+// Copyright 2017-2017 the openage authors. See copying.md for legal info.
+
+#include "../gamestate/player.h"
+#include "research.h"
+
+
+namespace openage {
+
+ResearchType::ResearchType(Player &owner)
+	:
+	owner{owner} {
+}
+
+std::shared_ptr<Research> ResearchType::initialise() const {
+	return std::make_shared<Research>(this);
+}
+
+Research::Research(const ResearchType *type)
+	:
+	type{type},
+	active_count{0},
+	completed_count{0} {
+}
+
+void Research::started() {
+	this->active_count += 1;
+}
+
+void Research::stopped() {
+	this->active_count -= 1;
+}
+
+void Research::completed() {
+	this->active_count -= 1;
+	this->completed_count += 1;
+}
+
+bool Research::can_start() const {
+	return this->active_count + this->completed_count < this->type->get_max_repeats();
+}
+
+bool Research::is_researched() const {
+	return this->completed_count == this->type->get_max_repeats();
+}
+
+void Research::apply() {
+	// apply the patch
+	this->type->apply();
+
+	// perform category based actions
+	if (type->category() == research_category::age_advance) {
+		type->owner.advance_age();
+
+	} else if (type->category() == research_category::age_advance) {
```

Comment:
we can do the research type-specific directly with nyan by testing the patch hierarchy:
```
nyan::Object research = ...;  // the research nyan object that now shall apply
// apply the research patch
nyan::Transaction tx = database.transaction();
for (auto &patch : research.get<nyan::Patch>("patches")) {
    tx.add(research);
}
tx.apply();

if (research.extends("openage.research.AgeAdvance")) {
    // do things for age advancements that nyan didn't patch (i don't know what that would be, maybe gui updates?)
}
```

Reply:
(I'm back)

EDIT: This was actually a typo: `advance_age` is handled at the first if, this was supposed to be `generic`, thinks that couldn't be handled just with nyan (eg. research that allows to see the enemy line of sight). 
However my following explanation for the `advance_age` still stands

The code you posted will be 
inside NyanResearchType::apply -> 
which ovverides ResearchType::apply -> 
which is called inside Reaseach::apply -> 
which is called from the ReaseachAction::update

Notes:
  ResearchType immutable (contains the name, cost, patch ...)
  Reaseach muttable (is active, is completed ...)

Based on that when the gui will have to draw the progress bar at the top of the screen it need only `action.research.get_categroy() == advance_age` and `action.timer.get_progress()`

No need for the gui to dive into the nyan hierarchy...

Reply:
Welcome back :)
Ah now I get it. Yes, we need to implement some research types in the engine, e.g. the line of sight. This would be done by specifying an object layout in nyan, which is then used by the c++-part to evaluate the data changes, e.g. the line of sight or whatever is needed.

The call hierarchy is correct I'd say. It would be a nice abstraction over the nyan backend.
I also like the difference between `Research` and `ResearchType` (i.e. ongoing vs its type).

To make the gui nyan-independent should be a high goal :D 

This wrapping of nyan-types is a great idea to write the interface to nyan: Then we'd have C++-classes that do the intended access to nyan objects (=> they're tightly coupled) and the rest-code can just use the C++-interface. The way it should be done.